[MTIA] Allow users who know what they are doing to ignore all device mismatches in tracing and take a preferred device.

patrick-toulme opened the pull request:
Summary:
Device mismatches in tracing can most often be ignored. These are only logical mismatches not physical. 

Take any intermediate computation, and that computation will not actually materialize in a compiled binary execution. So a device mismatch in the middle of the program is not real. The runtime will never materialize those tensors on CPU device during the execution, as they are temporary allocations. 

If a user knows his tensors at graph input are all on the correct device, then he can ignore all tracing errors.

Users who know what they are doing should have an escape hatch to ignore any device mismatch in tracing.

Users can set 
```
  torch._functorch.config.fake_tensor_prefer_device_type = 'mtia'
```
to forcefully override any mismatch and prefer the non cpu device. This unblocks vLLM graph mode for MTIA.

Test Plan:
Added two unit tests.

Rollback Plan:

Differential Revision: D79698438

patrick-toulme commented:
Tests passed local but failed on PR. Will debug tests and push again

jansel commented on torch/_functorch/config.py:
```diff
@@ -281,6 +281,16 @@ def remote_autograd_cache_default() -> Optional[bool]:
+# physically materialize in the binary execution, so the device mismatch
+# is not a real runtime concern. Enabling this allows the compiler to
+# proceed with compilation by choosing the non-CPU device for consistency.
+fake_tensor_prefer_non_cpu_device = False
```
Maybe this should be:
```py
fake_tensor_prefer_device_type : Optional[str] = None
```
So it can be a bit more general (rather than hardcoding non-cpu)

patrick-toulme replied:
@jansel  Done. Changed to as you described

patrick-toulme committed:
[MTIA] Allow users who know what they are doing to ignore all device mismatches in tracing and take the non CPU device. (#159931)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/159931

Device mismatches in tracing can most often be ignored. These are only logical mismatches not physical.

Take any intermediate computation, and that computation will not actually materialize in a compiled binary execution. So a device mismatch in the middle of the program is not real. The runtime will never materialize those tensors on CPU device during the execution, as they are temporary allocations.

If a user knows his tensors at graph input are all on the correct device, then he can ignore all tracing errors.

Users who know what they are doing should have an escape hatch to ignore any device mismatch in tracing.

Users can set
```
        torch._functorch.config.fake_tensor_prefer_device_type = 'mtia'
```
to forcefully override any mismatch and prefer the non cpu device. This unblocks vLLM graph mode for MTIA.

Test Plan:
Added two unit tests.

Rollback Plan:

Differential Revision: D79698438

patrick-toulme commented:
@pytorchbot merge

patrick-toulme commented:
@pytorchbot merge
